### PR TITLE
[WIP] Introduce `FlatParamHandle` and initial non-recursive path

### PR DESCRIPTION
**Pull request by awgu**
## Overview

1. This PR introduces `FlatParamHandle` to enable non-recursive FSDP wrapping. The class absorbs the unflattening/flattening logic from `FlattenParamsWrapper` but does not require wrapping a particular `nn.Module`.
2. This PR provides an initial non-recursive wrapping path with a naive `FlatParameter`/`FlatParamHandle` construction that iterates in `model.parameters()` order and uses a fixed bucket threshold. This is to ensure that the forward/backward code is adapted correctly for the non-recursive path, and we can focus on plugging the `FlatParameter` construction in the future.

### Introducing `FlatParamHandle`
There is flexibility in the design space for how to allocate attributes and methods to `FlatParameter` versus a wrapping class like `FlatParamHandle` or `FlattenParamsWrapper`. Several points in the design space provide the same functionality, so deciding on an allocation is arguably stylistic, though then preference should be given to cleaner designs.

The forefront consideration is that a `FlatParameter`'s metadata should be initialized once, while its data may be reloaded via checkpointing. This motivates decoupling the metadata initialization from the `FlatParameter` constructor, which should instead only handle the parameter data. Thus, we have both a `FlatParamHandle` managing a `FlatParameter` and the `FlatParameter` itself.
```
class FlatParamHandle:
    def __init__(self, module: nn.Module, params: Sequence[nn.Parameter]):
        # Calls `_init_flat_param()`
    def _init_flat_param(self, module: nn.Module, params: Sequence[nn.Parameter]):
	    # Calls `flatten_params()` and initializes metadata
    @staticmethod
	def flatten_params(params: Sequence[torch.Tensor], requires_grad: bool) -> FlatParameter:
		# Also may be used for checkpoint reloading
class FlatParameter(nn.Parameter):
	# Constructor is not overridden
```
Under this separation with `FlatParameter` as solely as a data container, we keep methods manipulating `FlatParameter` on the `FlatParamHandle`. Because `FlatParameter`'s constructor is not overridden, we should be able to replace it with another tensor type e.g. `ShardedTensor` with minimal changes.

### Compatibility with `FlattenParamsWrapper`
To ensure backward compatibility, `FlattenParamsWrapper` now holds a `FlatParamHandle`. Existing logic from `FlattenParamsWrapper` simply routes to the handle now.

A `FullyShardedDataParallel` instance holds references to all of its handles.
- For the recursive-wrapping paradigm, there is at most one handle, which is from its `FlattenParamsWrapper` if it manages parameters.
- For the non-recursive wrapping paradigm, there may be multiple handles, all owned by the single (root) `FullyShardedDataParallel` instance.

### `FlatParameter` Construction
In the existing implementation, a `FlatParameter`'s metadata was partially initialized in its constructor (e.g. `_param_numels`, `_param_shapes`) and partially initialized by the owning `FlattenParamsWrapper` (e.g. `_param_infos`, `_shared_param_infos`). The latter part was needed due to requiring module information. With this PR, the metadata initialization is consolidated in `FlatParamHandle`.
- During model construction, a `FlatParameter` should be initialized via the handle constructor`FlatParamHandle(module, params)`.
- During sharded checkpoint loading, a `FlatParameter` should be initialized via the static method `FlatParamHandle.flatten_params(new_params)`.
    - The checkpointing implementation is responsible for checking that `new_params` used to construct the `FlatParameter` data to load is consistent with the existing `FlatParameter`'s metadata.

These are the only two cases for `FlatParameter` construction right now, so there is no real functionality regression by not recomputing some of the metadata in the `FlatParameter` constructor. The `nn.Module.state_dict()` is implemented using in-place `copy_()`, so the new loaded `FlatParameter`'s metadata *should* match the existing `FlatParameter`'s metadata for correctness anyway. (I.e. we do not support a usage where we reload a `FlatParameter` with differing metadata into an existing `FlatParameter`.)

### BC Breaking
- `ShardMetadata` -> `FlatParamShardMetadata` to avoid name conflict with `ShardedTensor`
    - `param_names` prefixed only by immediate parent module's name -> `param_names` fully-prefixed starting from `module` passed into `FlattenParamsWrapper`/`FlatParamHandle`
        - This guarantees uniqueness within the `FullyShardedDataParallel` instance
        - For example, before, `Model(Sequential(Linear, Sequential(Linear)))` with `always_wrap_policy` will have have `0.weight` and `0.bias` for both `Linear`s
    - `metadata()` -> removed (unused)
- `FlatParameter` attributes
    - `_param_numels` -> `_numels`
    - `_param_shapes` -> `_shapes`
    - `full_numel` -> `_unsharded_size.numel()`
    - `_param_indice_in_shard` -> `_shard_param_indices`
    - `_sharded_param_offsets` -> `_shard_param_offsets`
    - `num_padded` -> `_shard_numel_padded`
    - `param_offsets` -> not saved; directly constructed in `_get_flat_param_offsets()` and used once
- `FlattenParamsWrapper` `param_list` argument -> `params` for consistency with `FlatParameter`

### Naming

We need to decide on a name for this "non-recursive wrapping" code path. Currently, I have a `_ExecOrderPolicy` class that is used to designate that code path. We should settle on something and stay consistent.

### Backward Compatibility
- [] Run existing unit tests
- [] Run QPS benchmark
- [] Run accuracy benchmark
- [] Run large-model to check memory management
### Non-Recursive Wrapping
- [] Add training unit tests (forward/backward/optimizer step) covering Cartesian product of all features
    - [] Without shared parameters
    - [] With shared parameters
- [] Add model checkpointing unit tests
    - [] Full `state_dict()`
    - [] Local `state_dict()`
   - [] Sharded `state_dict()`
- [] Add optimizer checkpointing unit tests
- [] Run QPS benchmark
- [] Run accuracy benchmark

## Follow-Ups
- The current `FlatParameter`'s `data` represents either the sharded unflattened parameter, unsharded unflattened parameter, or reduced-precision sharded unflattened parameter, depending dynamically on the runtime context. When its `data` represents one quantity, the other quantities are still saved as attributes on the `FlatParameter` (e.g. `_local_shard`, `_full_param_padded`, `_mp_shard`). `FullyShardedDataParallel` directly manipulates the `data`.
We should investigate the tradeoffs of having those attributes on the `FlatParameter` versus moving them to the `FlatParamHandle`. The motivation for the latter is to define a clean interface for `FullyShardedDataParallel` to manage parameter data in preparation for generalizing to multiple parameter groups, to managing non-`FlatParameter`s, and to supporting non-CUDA devices. (More explicitly, `FullyShardedDataParallel`'s parameter *variables* would be set to different `Tensor` variables, none of which own another, instead of `FullyShardedDataParallel`'s parameter variables' *data* being set to different `Tensor` variables, all owned by the `FlatParameter`, and the data management would be folded into handle, hidden from `FullyShardedDataParallel`.)
- We should investigate if we can coalesce the remaining logic in `FlattenParamsWrapper` into `FullyShardedDataParallel` and remove `FlattenParamsWrapper`.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2262,7 +2365,83 @@ def _load_sharded_state_dict(
...
+                    args = args[0]
+                    kwargs = kwargs[0]
+                    if self._mixed_precision_enabled_for_params():
+                        input_dtype = self.mixed_precision.param_dtype
+                        args, kwargs = self._cast_fp_inputs_to_precision(
+                            input_dtype, *args, **kwargs,
+                        )
+                handles = self._module_to_handles[module]
+                params_to_reshard = []
+                # TODO (awgu): Replace the below logic to check if the handles'
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2403]: @LinjianMa This is the point that determines if we have good memory savings by resharding in a timely manner. We want to create a graph structure that tracks the dependencies so that we know exactly which parameters we can reshard here.

I think we can just only free in the "pre-forward" (before forward computation) and not free in the "post-forward" (after forward computation) at all. This means that the last module will not free, but this was the intention behind setting the root's `reshard_after_forward=False` to begin with. The last module will most likely participate in gradient computation immediately, so there is no need to reshard and re-all-gather.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1079,13 +1095,96 @@ def _check_wrapped(cls, begin_module, check_fn, err_fn):
...
+            params: A list of all :class:`FlatParameter` s.
+            _module_to_forward: A mapping from original module to the module's
+                unpatched ``forward()`` method.
+        """
+        # For the non-recursive wrapping, we set TODO
+        self._fsdp_wrapped_module = root_module
+        self._module_to_handles: Dict[nn.Module, List[FlatParamHandle]] = \
+            collections.defaultdict(list)
+
+        # TODO (awgu): For now, we assume a fixed bucket size and follow
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1153]: @LinjianMa  I am doing the handle creation in the FSDP constructor right now. We will need to move this to after the first iteration (probably in the post-backward callback).

This just goes in `.parameters()` order and buckets according to some threshold. This is good for testing since I can make the bucket threshold small to force many handles to be created.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1079,13 +1095,96 @@ def _check_wrapped(cls, begin_module, check_fn, err_fn):
...
+        Defines:
+            _fsdp_wrapped_module
+            _module_to_handles: A mapping from original module to the handles
+                that contain some part of its parameters.
+            params: A list of all :class:`FlatParameter` s.
+            _module_to_forward: A mapping from original module to the module's
+                unpatched ``forward()`` method.
+        """
+        # For the non-recursive wrapping, we set TODO
+        self._fsdp_wrapped_module = root_module
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1149]: Maybe it's better to add a comment in the constructor: for the nonrecursive wrap policy it represents the original module, otherwise its a FlattenParameterWrapper

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -3299,6 +3508,9 @@ def _free_full_params(self, params: Optional[List[FlatParameter]] = None) -> Non
             params = self.params
         current_stream = torch.cuda.current_stream()
         for p in params:
+            # TODO (awgu): do we want this here?
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3514]: I think so

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -2262,7 +2365,83 @@ def _load_sharded_state_dict(
...
+        return patched_forward
+
+    def _unpatch_forwards(self):
+        """Unpatches ``module.forward()`` to remove additional FSDP logic."""
+        for module, forward in self._module_to_forward.items():
+            module.forward = forward
+
     def forward(self, *args: Any, **kwargs: Any) -> Any:
+        if self._use_exec_order_policy:
+            assert self._is_root is None or self._is_root
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2443]: and rather than or?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -2262,7 +2365,83 @@ def _load_sharded_state_dict(
...
+                if self._is_root:
+                    args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
+                    args = args[0]
+                    kwargs = kwargs[0]
+                    if self._mixed_precision_enabled_for_params():
+                        input_dtype = self.mixed_precision.param_dtype
+                        args, kwargs = self._cast_fp_inputs_to_precision(
+                            input_dtype, *args, **kwargs,
+                        )
+                handles = self._module_to_handles[module]
```
Comment (LinjianMa) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2401]: I am sometimes confused with handles and self._handles. I guess for this we can just use self._module_to_handles[module] throughout within the patch_forward function, this would be clearer to me

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2403]: I think post-forward patching is necessary, one example is if we have a forward F1 with parameter P1 and its children forwards F2 and F3 with parameters P2 and P3, with the following structure
[P1, [P2], P1, [P3]], then before the forward of F2, P1 will be resharded. After the forward of F2, we need to rebuild full params of P1 to make sure the forward of F1 is correct.

**Review by LinjianMa — COMMENTED** [review]
The logic and general structure makes sense to me, thank you for all the work! Below are my comments based on the initial review,

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 5 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2443]: If `_lazy_init()` has not been called yet (i.e. first iteration), then `self._is_root` is `None`.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1149]: 😆 sorry I was still experimenting with how I wanted to do this (hence the `TODO` in the comment). I will definitely clean this up.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 6 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2401]: I will try to make it clearer. In general, `self._handles` will always represent *all* handles, while a local variable `handles` may be specific to that context (e.g. a module's `forward()`).
